apps/cloudflare/src/inngest/middlewares/rate-limit-middleware.ts

```diff
@@ -0,0 +1,54 @@
+import { InngestMiddleware, RetryAfterError } from 'inngest';
+import { CloudflareError } from '@/connectors/commons/error';
+
+/**
+ * This middleware, `rateLimitMiddleware`, is designed for use with the Inngest serverless framework.
+ * It aims to handle rate limiting scenarios encountered when interacting with external SaaS APIs.
+ * The middleware checks for specific errors (instances of CloudflareError) that indicate a rate limit has been reached,
+ * and it responds by creating a RetryAfterError. This error includes the retry time based on the 'Retry-After' header
+ * provided by the SaaS service, enabling the function to delay its next execution attempt accordingly.
+ *
+ * Key Features:
+ * - Intercepts function output to check for rate limit errors.
+ * - Handles CloudflareError, specifically looking for a 'Retry-After' header in the error response.
+ * - Generates a RetryAfterError to reschedule the function run, preventing immediate retries that could violate the SaaS's rate limits.
+ *
+ * Note: This is a generic middleware template and might require adjustments to fit specific SaaS APIs' error handling and rate limiting schemes.
+ */
+export const rateLimitMiddleware = new InngestMiddleware({
+  name: 'rate-limit',
+  init: () => {
+    return {
+      onFunctionRun: ({ fn }) => {
+        return {
+          transformOutput: (ctx) => {
+            const {
+              result: { error, ...result },
+              ...context
+            } = ctx;
+            const retryAfter =
+              error instanceof CloudflareError && error.response?.headers.get('Retry-After');
+
+            if (!retryAfter) {
+              return;
+            }
+
+            return {
+              ...context,
+              result: {
+                ...result,
+                error: new RetryAfterError(
+                  `Cloudflare rate limit reached by '${fn.name}'`,
+                  retryAfter,
+                  {
+                    cause: error,
+                  }
+                ),
+              },
+            };
```

Comment:
According to cloudflare doc, this should be our implementation:
```suggestion
            if (error.response?.status !== 429) {
              return;
            }

            return {
              ...context,
              result: {
                ...result,
                error: new RetryAfterError(
                  `Cloudflare rate limit reached by '${fn.name}'`,
                 addMinutes(new Date(), 5),
                  {
                    cause: error,
                  }
                ),
              },
            };
```

Reply:
Changes Done